### ISSUE TITLE
Fix nested if statements

### ISSUE THREAD

**Issue by nickwanninger** [nickwanninger/geode#4]
In gitlab by @nickwanninger on Jun 24, 2018, 22:56

the code
```
func main(int argc) byte {
	int a := 3;
	if argc = 2 {
		printf("Is GT Two\n");
		if a = 3 {
			printf("'a' is 3\n");
		}
                # merge_8
	}
	printf("%d\n", a);
	return 0;
}
```

Fails to compile. It produces the following llvm for main
```llvm
define i8 @main(i64 %argc) {
entry_1:
        %0 = alloca i64
        store i64 %argc, i64* %0
        %a = alloca i64
        store i64 3, i64* %a
        %1 = load i64, i64* %0
        %2 = icmp eq i64 %1, 2
        %3 = icmp eq i1 true, %2
        br i1 %3, label %then_2, label %else_3
then_2:
        %4 = call i64 (i8*, ...) @printf(i8* getelementptr ([11 x i8], [11 x i8]* @.str_5, i32 0, i32 0))
        %5 = load i64, i64* %a
        %6 = icmp eq i64 %5, 3
        %7 = icmp eq i1 true, %6
        br i1 %7, label %then_6, label %else_7
else_3:
        br label %merge_4
merge_4:
        %8 = load i64, i64* %a
        %9 = call i64 (i8*, ...) @printf(i8* getelementptr ([4 x i8], [4 x i8]* @.str_10, i32 0, i32 0), i64 %8)
        %10 = trunc i64 0 to i8
        ret i8 %10
then_6:
        %11 = call i64 (i8*, ...) @printf(i8* getelementptr ([10 x i8], [10 x i8]* @.str_9, i32 0, i32 0))
        br label %merge_8
else_7:
        br label %merge_8
merge_8:
        %!s(<nil>)
}
```

Notice the `nil` at the bottom? That is because block pushing and popping fails to work correctly and merge_8 is technically the space after the nested if statement

**Comment by nickwanninger** [comment]
In gitlab by @nickwanninger on Jun 24, 2018, 23:05

On initial inspection, it might be something to do with the block stack system in place

**Comment by nickwanninger** [comment]
In gitlab by @nickwanninger on Jun 24, 2018, 23:17

I rebuilt the block stack to use an array instead of a fancy linked list. Still have the problem

**Comment by nickwanninger** [comment]
In gitlab by @nickwanninger on Jun 24, 2018, 23:35

I managed to get that code to compile into 
```llvm
define i8 @main(i64 %argc) {
entry_1:
        %0 = alloca i64
        store i64 %argc, i64* %0
        %a = alloca i64
        store i64 3, i64* %a
        %1 = load i64, i64* %0
        %2 = icmp eq i64 %1, 2
        %3 = icmp eq i1 true, %2
        br i1 %3, label %if_0_then_2, label %if_0_else_8
if_0_then_2:
        %4 = call i64 (i8*, ...) @printf(i8* getelementptr ([11 x i8], [11 x i8]* @.str_3, i32 0, i32 0))
        %5 = load i64, i64* %a
        %6 = icmp eq i64 %5, 3
        %7 = icmp eq i1 true, %6
        br i1 %7, label %if_1_then_4, label %if_1_else_6
if_1_then_4:
        %8 = call i64 (i8*, ...) @printf(i8* getelementptr ([10 x i8], [10 x i8]* @.str_5, i32 0, i32 0))
        br label %if_1_merge_7
if_1_else_6:
        br label %if_1_merge_7
if_1_merge_7:
        %!s(<nil>)
if_0_else_8:
        br label %if_0_merge_9
if_0_merge_9:
        %9 = load i64, i64* %a
        %10 = call i64 (i8*, ...) @printf(i8* getelementptr ([4 x i8], [4 x i8]* @.str_10, i32 0, i32 0), i64 %9)
        %11 = trunc i64 0 to i8
        ret i8 %11
}
```

Which is correct. Except the internal merge is goofy. I need to figure out how to get that to then branch to the `if_0_merge_9` block.

**Comment by nickwanninger** [comment]
In gitlab by @nickwanninger on Jun 24, 2018, 23:50

Problem solved. Turns out, it is important to know the block after being generated. So I added a return statement to the blockNode's codegen function that returns the current Block at time of codegen. This way, in the ifNode's codegen, I can ensure they contain branches back to the if statement's merge.

This program works now :)
```
func main(int argc) byte {
	int a := 1;
	if argc >= 2 {
		a <- 2;
		# then 0
		printf("at least 2 args\n");
		if argc >= 3 {
			a <- 3;
			printf("at least 3 args\n");
			if argc >= 4 {
				a <- 4;
				printf("at least 4 args\n");
				if argc >= 5 {
					a <- 5;
					printf("at least 5 args\n");
					if argc >= 6 {
						a <- 6;
						printf("at least 6 args\n");
					} else {printf("5's it\n");}
				} else {printf("4's it\n");}
			} else {printf("3's it\n");}
		} else {printf("2's it\n");}
		# merge 1
	} else {printf("1 arg :(\n");}
	# merge 0
	printf("%d\n", a);
	return 0;
}
```

**Comment by nickwanninger** [comment]
In gitlab by @nickwanninger on Jun 24, 2018, 23:52

Fixed in 05cab25adc306d2f1f7d3f6093b6d6f785c9367f

**Comment by nickwanninger** [comment]
In gitlab by @nickwanninger on Jun 24, 2018, 23:52

closed
